odbc/src/handles.rs

```diff
@@ -0,0 +1,202 @@
+use std::{
+    collections::HashSet,
+    sync::{Mutex, RwLock},
+};
+
+use odbc_sys::Integer;
+
+#[derive(Debug)]
+pub struct EnvHandle {
+    pub env: RwLock<Env>,
+}
+
+impl EnvHandle {
+    pub fn new() -> Self {
+        Self {
+            env: RwLock::new(Env {
+                _attributes: EnvAttributes::default(),
+                state: EnvState::Unallocated,
+                connections: HashSet::new(),
+            }),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct Env {
+    // attributes for this Env
+    pub _attributes: EnvAttributes,
+    // state of this Env
+    pub state: EnvState,
+    pub connections: HashSet<*mut ConnectionHandle>,
+}
+
+#[derive(Debug)]
+pub struct EnvAttributes {
+    pub odbc_ver: Integer,
+    // TODO: incomplete
+}
+
+impl Default for EnvAttributes {
+    fn default() -> Self {
+        Self { odbc_ver: 3 }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum EnvState {
+    Unallocated,
+    Allocated,
+}
+
+#[derive(Debug)]
+pub struct ConnectionHandle {
+    pub connection: RwLock<Connection>,
+}
+
+#[derive(Debug)]
+pub struct Connection {
+    // Pointer to the Env from which
+    // this Connection was allocated
+    pub env: *mut EnvHandle,
+    // all the possible Connection settings
+    pub _attributes: ConnectionAttributes,
+    // state of this connection
+    pub state: ConnectionState,
+    // MongoDB Client for issuing commands
+    // pub client: Option<MongoClient>,
+    // all Descriptors attached to this Connection
+    pub _descriptors: HashSet<*mut DescriptorHandle>,
+    // all Statements allocated from this Connection
+    pub statements: HashSet<*mut StatementHandle>,
+}
+
+#[derive(Debug)]
+pub struct ConnectionAttributes {}
+
+impl Default for ConnectionAttributes {
+    fn default() -> Self {
+        Self {}
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum ConnectionState {
+    _UnallocatedEnvUnallocatedConnection,
+    AllocatedEnvUnallocatedConnection,
+    AllocatedEnvAllocatedConnection,
+    _ConnectionFunctionNeedsDataEnv,
+    _Connected,
+    _StatementAllocated,
+    _TransactionInProgress,
+}
+
+impl ConnectionHandle {
+    pub fn new(env: *mut EnvHandle) -> Self {
+        Self {
+            connection: RwLock::new(Connection::new(env)),
+        }
+    }
+}
+
+impl Connection {
+    pub fn new(env: *mut EnvHandle) -> Self {
+        Self {
+            env,
+            _attributes: ConnectionAttributes::default(),
+            state: ConnectionState::AllocatedEnvUnallocatedConnection,
+            _descriptors: HashSet::new(),
+            statements: HashSet::new(),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct StatementHandle {
+    pub stmt: Mutex<Statement>,
```

Comment:
I see that you called this out in the design doc, but asking here instead because this is where other discussion threads already exist:

Why does this need to be a mutex? I'm not sure I understand why a `Cursor` is different wrt what kind of locking we need

Reply:
Reading a cursor is a write, also. Maybe this is unnecessary, just trying to be safe as possible.

Reply:
Actually, I think it makes sense to change to `RwLock` and we'll just have to use `write` to access the cursor.

Reply:
yeah, I think the `Cursor` API already prevents this -- you need a mutable borrow to iterate the cursor!